src/riak_kv_w1c_worker.erl

```diff
@@ -354,3 +349,82 @@ erase_request_record(ReqId, #state{entries=Entries} = State) ->
         OldValue ->
             {OldValue, State#state{entries=?DICT_TYPE:erase(ReqId, Entries)}}
     end.
+
+%%%% Functions to handle asynchronous puts
+
+%% Invoked by put/2 to turn the async request into a synchronous call
+synchronize_put({ok, {ReqId, Worker}}, Options) ->
+    Timeout = case proplists:get_value(timeout, Options, ?DEFAULT_TIMEOUT) of
+                  N when is_integer(N) andalso N > 0 ->
+                      N;
+                  _ ->
+                      ?DEFAULT_TIMEOUT
+              end,
+    %% this is not direct gen_server:call because the process that
+    %% replies is not the one that received the cast
+    receive
+        {'DOWN', ReqId, process, _Pid, _Reason} ->
+            {error, riak_kv_w1c_server_crashed};
+        {ReqId, Response} ->
+            erlang:demonitor(ReqId, [flush]),
+            Response
+    after Timeout ->
+            gen_server:cast(Worker, {cancel, ReqId}),
+            receive
+                {'DOWN', ReqId, process, _Pid, _Reason} ->
+                    {error, riak_kv_w1c_server_crashed};
+                {ReqId, Response} ->
+                    erlang:demonitor(ReqId, [flush]),
+                    Response
+            end
+    end;
+synchronize_put(Error, _Options) ->
+    Error.
+
+%% Invoked by async_put_reply/2 to wait for all responses
+async_put_reply_loop([], Responses, _Timeout) ->
+    Responses;
+async_put_reply_loop(ReqIds, Responses, Timeout) ->
+    receive
+        {'DOWN', ReqId, process, _Pid, _Reason} when is_reference(ReqId) ->
```

Comment:
Notice that it's pattern matching on the reference given back to us by monitor. Does that alleviate your concerns?

Sent from my iPhone

> On Nov 4, 2015, at 2:24 PM, Micah <notifications@github.com> wrote:
> 
> In src/riak_kv_w1c_worker.erl:
> 
> > +                {'DOWN', ReqId, process, _Pid, _Reason} ->
> > +                    {error, riak_kv_w1c_server_crashed};
> > +                {ReqId, Response} ->
> > +                    erlang:demonitor(ReqId, [flush]),
> > +                    Response
> > +            end
> > +    end;
> > +synchronize_put(Error, _Options) ->
> > +    Error.
> > +
> > +%% Invoked by async_put_reply/2 to wait for all responses
> > +async_put_reply_loop([], Responses, _Timeout) ->
> > +    Responses;
> > +async_put_reply_loop(ReqIds, Responses, Timeout) ->
> > +    receive
> > +        {'DOWN', ReqId, process, _Pid, _Reason} when is_reference(ReqId) ->
> This concerns me a bit. If the calling process is monitoring anything other than the put worker, and that other goes down after the async_put_reply_loop starts (or even before it's able to receive that down message), it gets swallowed here, enduing up with an incorrect 'w1c_server crashed' error. This is repeated in the aysnc_put_cleanup function.
> 
> —
> Reply to this email directly or view it on GitHub.
>